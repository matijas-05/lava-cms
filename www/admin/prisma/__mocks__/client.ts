import { afterEach, beforeEach, vi } from "vitest";
import { mockDeep, mockClear } from "vitest-mock-extended";
import type { PrismaClient } from "@prisma/client";

export const prisma = mockDeep<PrismaClient>();

beforeEach(() => {
	// Bypass the privateAuth middleware
	prisma.config.count.mockResolvedValue(0);
	vi.mock("next/headers", () => ({
		cookies: vi.fn(() => ({
			get: vi.fn(),
		})),
	}));
	vi.mock("@admin/src/utils/server", () => ({
		url: vi.fn(() => "http://localhost:3001"),
	}));
});
afterEach(() => {
	mockClear(prisma);
	vi.clearAllMocks();
});
