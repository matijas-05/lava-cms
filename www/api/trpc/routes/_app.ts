import { router } from "@api/trpc/trpc";
import { authRouter } from "./auth/_auth";

export const appRouter = router({
	auth: authRouter,
});
export const caller = appRouter.createCaller({});

export type AppRouter = typeof appRouter;
