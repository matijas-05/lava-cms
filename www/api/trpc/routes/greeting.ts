import { publicProcedure } from "../trpc";
import { z } from "zod";

export const greeting = publicProcedure
	.input(
		z.object({
			name: z.string(),
		})
	)
	.query(({ input }) => {
		return {
			greeting: "Hello " + input?.name,
		};
	});
