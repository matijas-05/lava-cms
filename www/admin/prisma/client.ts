import { PrismaClient } from "@prisma/client";
import { env } from "@/src/env/server.mjs";

const globalForPrisma = global as unknown as { prisma: PrismaClient };

export const prisma =
	globalForPrisma.prisma ||
	new PrismaClient({
		log: ["query", "info", "warn", "error"],
	});

if (env.NODE_ENV !== "production") globalForPrisma.prisma = prisma;
