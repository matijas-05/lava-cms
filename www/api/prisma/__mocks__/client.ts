import { beforeEach } from "vitest";
import { mockDeep, mockReset } from "vitest-mock-extended";
import { type PrismaClient } from "@prisma/client";

export const prisma = mockDeep<PrismaClient>();

beforeEach(() => {
	mockReset(prisma);
});
