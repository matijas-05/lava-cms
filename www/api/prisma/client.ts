import { PrismaClient } from "@prisma/client";
import { env } from "@api/env/server";

const globalForPrisma = global as unknown as { prisma: PrismaClient };

export const prisma =
	globalForPrisma.prisma ||
	new PrismaClient({
		errorFormat: "colorless",
		log: ["query", "info", "warn", "error"],
	});

if (env.NODE_ENV !== "production") globalForPrisma.prisma = prisma;
