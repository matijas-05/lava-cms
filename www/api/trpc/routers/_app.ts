import { router } from "../trpc";
import { greeting } from "./greeting";
import { getUser } from "./getUser";

export const appRouter = router({
	greeting,
	getUser
});

export type AppRouter = typeof appRouter;
