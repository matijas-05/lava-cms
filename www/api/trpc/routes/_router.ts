import { router } from "../trpc";
import { greeting } from "./greeting";
import { firstTime } from "./auth/firstTime";
import { signUp } from "./auth/signUp";
import { signIn } from "./auth/signIn";

const auth = {
	firstTime,
	signUp,
	/**
	 * @returns The user's ID
	 */
	signIn,
};

export const appRouter = router({
	greeting,
	...auth,
});
export const caller = appRouter.createCaller({});

export type AppRouter = typeof appRouter;
