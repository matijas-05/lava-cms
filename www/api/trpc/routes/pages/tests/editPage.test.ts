import { expect, it, vi } from "vitest";
import { Prisma } from "@prisma/client";
import { prisma } from "@api/prisma/__mocks__/client";
import { caller } from "@api/trpc/routes/_app";
import type { TRPC_ERROR_CODE_KEY } from "@trpc/server/rpc";

vi.mock("@api/prisma/client");

const ID = "clgaxfcla0017qnotc4a9aemf";
const NEW_NAME = "New Name";
const NEW_SLUG = "new-name";

it("updates the page's url and its children's urls", async () => {
	prisma.page.findFirst.mockResolvedValue({
		id: ID,
		name: "Old Name",
		url: "/old-name",
		order: 0,
		parent_id: null,
	});

	await caller.pages.editPage({
		id: ID,
		newName: NEW_NAME,
		newUrl: "/" + NEW_SLUG,
	});

	const updatePageCall = prisma.page.update.mock.calls[0];
	expect(updatePageCall).toBeDefined();

	const updatePageArgs = updatePageCall![0];
	expect(updatePageArgs).toEqual({
		where: { id: ID },
		data: {
			name: NEW_NAME,
			url: "/" + NEW_SLUG,
		},
	});

	const updateChildrenCall = prisma.$executeRaw.mock.calls[0];
	expect(updateChildrenCall).toBeDefined();
	expect(updateChildrenCall).toMatchSnapshot();
});

it("throws a trpc 404 'NOT_FOUND' error if the page doesn't exist", async () => {
	prisma.page.findFirst.mockResolvedValue(null);

	await expect(
		caller.pages.editPage({
			id: ID,
			newName: NEW_NAME,
			newUrl: "/" + NEW_SLUG,
		})
	).rejects.toThrowError("NOT_FOUND" as TRPC_ERROR_CODE_KEY);
});

it("throws a trpc 409 'CONFLICT' error if the new url is already taken", async () => {
	prisma.page.findFirst.mockResolvedValue({
		id: ID,
		name: NEW_NAME,
		url: "/" + NEW_SLUG,
		order: 0,
		parent_id: null,
	});

	prisma.$transaction.mockRejectedValue(
		new Prisma.PrismaClientKnownRequestError(
			"Unique constraint failed on the fields: (`url`)",
			{
				clientVersion: "mocked",
				code: "P2002",
				meta: {
					target: ["url"],
				},
			}
		)
	);

	await expect(
		caller.pages.editPage({
			id: ID,
			newName: NEW_NAME,
			newUrl: "/" + NEW_SLUG,
		})
	).rejects.toThrowError("CONFLICT" as TRPC_ERROR_CODE_KEY);
});

it("throws a trpc 500 'INTERNAL_SERVER_ERROR' error if the database throws an unknown error", async () => {
	prisma.page.findFirst.mockResolvedValue({
		id: ID,
		name: NEW_NAME,
		url: "/" + NEW_SLUG,
		order: 0,
		parent_id: null,
	});

	prisma.$transaction.mockRejectedValue(new Error("Unknown error"));

	await expect(
		caller.pages.editPage({
			id: ID,
			newName: NEW_NAME,
			newUrl: "/" + NEW_SLUG,
		})
	).rejects.toThrowError("INTERNAL_SERVER_ERROR" as TRPC_ERROR_CODE_KEY);
});
