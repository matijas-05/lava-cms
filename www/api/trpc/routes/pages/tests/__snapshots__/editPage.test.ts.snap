// Vitest Snapshot v1

exports[`updates the page's url and its children's urls 1`] = `
[
  [
    "UPDATE frontend.page SET \\"url\\" = REPLACE(\\"url\\", ",
    " || '/', ",
    " || '/'), \\"last_update\\" = Now() WHERE \\"url\\" LIKE ",
    " || '/%';",
  ],
  "/old-name",
  "/new-name",
  "/old-name",
]
`;
